### PR TITLE
Start describing the state machines

### PR DESCRIPTION
**Pull request by pirapira**
This commit adds a file that describes state machines.  As a start, I tried to describe the ``HashTimeLockState``.

This is the first bite on https://github.com/raiden-network/raiden/issues/3160

[no ci integration]

### CODE REVIEW

File: docs/state_machines.rst [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+============
+
+The current Raiden prototype chose to use hierarchical state machines to represent the internal state of the node. The state of the whole node is a tree that can be serialized on and recovered from the disk. Moreover, whenever the state changes, the change is triggered by a state change object, which can also be serialized on the disk. A snapshot of the state and the subsequent state change objects are enough to compute the current state of the node. That's how a node recovers from crashes.
+
+Kinds of State Machines
+=======================
+
+The state of the whole node is represented as a ``ChainState`` object. The ``ChainState`` object contains many other different kinds of objects. This section describes them in a bottom-to-top manner.
+
+This section is very early in construction with the simplest state machines.
```
Comment (LefterisJP) [inline docs/state_machines.rst]: Can you perhaps rephrase this sentence? I don't really understand what you are trying to say.

File: docs/state_machines.rst [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+HashTimeLockState
+*****************
+
+A ``HashTimeLockState`` represents a hash time lock.
+
+Attributes of a HashTimeLockState
```
Comment (LefterisJP) [inline docs/state_machines.rst]: I would skip documenting the attributes of each state or state change. Simply point them to the documentation's docstring. No reason to repeat information here.

Comment (pirapira) [reply docs/state_machines.rst]: I've added comments to the source.  I cannot add a link in the spec right now until the comments are merged.

Comment (LefterisJP) [reply docs/state_machines.rst]: Source of the raiden client? Where?

Comment (pirapira) [reply docs/state_machines.rst]: Here https://github.com/raiden-network/raiden/pull/3177/files#diff-a6042df3a2b6bca66372154b6cceff89R1000 in this PR.

**Comment by pirapira** [comment]
@LefterisJP Merge?

**Pull request merged**